### DISCUSSION TITLE
Show us your Edrumulus implementation!

### DISCUSSION THREAD

**Discussion by corrados** [corrados/edrumulus#28] [Show and tell]
Did you build your own Edrumulus e-drum module? What are your experiences with Edrumulus? Please share it to the community!

Please post your report+pictures in this thread. Thanks :-)

**Comment by thijstriemstra** [comment]
I posted our [design on printables.com](https://www.printables.com/model/219389-esp32-midi-drum-module).

More info here: https://github.com/corrados/edrumulus/discussions/54

**Reply by corrados** [reply]
I'm really impressed by your case design. Looks much better than mine :-). Thank you for sharing it.

**Reply by thijstriemstra** [reply]
Thanks @corrados! We also decided to post some video of the machine in action: https://www.youtube.com/watch?v=YWE1LgwQbbQ

**Reply by corrados** [reply]
Great, thanks :-). Looks like your prototype is working to your satisfaction.

**Comment by corrados** [comment]
I have just uploaded a [Youtube video](https://youtu.be/tBlKvivwAE4) of my latest Edrumulus prototype (Prototype 3 (HD-1)) in action.

**Comment by ryanalexmartin** [comment]
I think edrumulus is the best open-source drum module available on GitHub, but I haven't exactly gotten it to work yet.  I've been tinkering with this repository for over a year, but I feel a bit nervous asking for help.  Perhaps edrumulus could benefit from a Discord community, where newcomers without electrical engineering or  DSP experience like myself could ask questions when we get stuck in the initial steps.  I've been stuck there for well over a year.

I've hand soldered the prototype 2 circuit on three different occasions:
![IMG_1180](https://github.com/corrados/edrumulus/assets/15975303/ee7793e4-dc2d-4cfe-a710-d6fdfbd73206)
![IMG_1179](https://github.com/corrados/edrumulus/assets/15975303/c09e2de0-4c78-417b-b949-468c98672c54)

Each time with different results.
<img width="609" alt="Screenshot 2023-08-14 at 10 17 37 PM" src="https://github.com/corrados/edrumulus/assets/15975303/c96b86d1-9fec-40ee-b2e9-064f6bf0cdcb">
Sometimes, when I compile and upload the code to my ESP32 dev board, it works, but usually I get some form of the above.  Constantly triggering notes, and the velocity value might fluctuate anywhere from a value of 0-1 to as high as 80-90.  I would like to ask questions to diagnose my circuit, or perhaps it's something wrong with small changes I made to the code.  For example, how can I configure edrumulus to test only one pad at a time?  These are the kinds of questions that a Discord (or similar) community would benefit.  

I'd love to contribute to this project, as well as its documentation, but even after digging around through the codebase for so long, I'm having a hard time getting my footing.

**Reply by ocimjr** [reply]
I'm a beginner in electronics as well. When I have multiple triggers, in my case, can be 2 thing:
1. I didn't connect to 3.3v pin
2. Something loose in the connection from one of circuits.

Hope it helps you

**Reply by corrados** [reply]
>  I've been tinkering with this repository for over a year, but I feel a bit nervous asking for help.

Please don't. Just ask your question. I'm happy to help you getting it working. But I would suggest that you open a new discussion thread for your question(s) so that other users can find it more easily. This discussion thread was intended for working prototypes.

**Reply by corrados** [reply]
I forgot to answer the questions you have already posted:

> I've hand soldered the prototype 2 circuit on three different occasions:

What do you mean by "three different occasions"?

I can see that you have soldered multiple input channels but you have connected only one to the ESP32 board. If you do not disable the other channels which are not connected, it will not work and you will get unwanted MIDI notes. Which leads me to your next question :-)

>  For example, how can I configure edrumulus to test only one pad at a time? 

This is simple. In the [these two code lines](https://github.com/corrados/edrumulus/blob/6fbba963e4070379057935972f63135191c61021/edrumulus_hardware.cpp#L229). only enable the very first entry which then only enables the snare input:

```
  static int analog_pins4[]         = { 36 };
  static int analog_pins_rimshot4[] = { 35 };
```

> I'd love to contribute to this project, as well as its documentation,

Yea, this is [already on the TODO list](https://github.com/corrados/edrumulus/blob/main/doc/TODO.md) but quite at the bottom of the list ;-). We already started to write a manual in [this file](https://github.com/corrados/edrumulus/blob/main/doc/manual.md) but it is already outdated and since it is not maintained, it is not referenced in the Edrumulus README.md at all. If you want to contribute, this file would be a very good start :-).

**Reply by thijstriemstra** [reply]
maybe you should start a new thread if you're having issues @ryanalexmartin, to keep this thread on topic.

**Comment by shepss123** [comment]
I started mine so far so good 
![20250125_144455](https://github.com/user-attachments/assets/286ee3fc-3fec-4ef1-83ed-34cb4b345117)
![20250207_074205](https://github.com/user-attachments/assets/c1827188-0a79-4b38-8eef-f99251ae7495)
![20250127_163029_001](https://github.com/user-attachments/assets/c9050809-0ec2-4f26-8c32-96fc2c92455e)
![20250127_163016](https://github.com/user-attachments/assets/debeeb31-979f-4bee-a4ca-7089d7d90ca3)

![20250207_080550](https://github.com/user-attachments/assets/162d9115-e5f1-407d-ba21-21edcba50ea3)
![20250207_080542](https://github.com/user-attachments/assets/65f5d78c-2d9c-412a-8847-342fe2d459ef)

![20250207_080605](https://github.com/user-attachments/assets/afaa7f7b-c375-42d6-aee5-cdad55c523ec)

**Reply by shepss123** [reply]
I repurposed some of the components I had from my previous arduino mega drum and I always wanted to use wood when building my midi projects

**Reply by corrados** [reply]
Looks nice. I hope that the finished device then works as expected :-).

**Comment by 3hhh** [comment]
Yes, that looks good.

I'm still with breadboards, but they appear to be stable for a long while now...
So imho soldering isn't really required apart from the cables unless you need gig-like stability.

![image](https://github.com/user-attachments/assets/f56b327a-e017-4d95-b41b-e76325c4ef45)

(Teensy and more breadboards are inside that plastic box. The jumper cables at the top are soldered to standard stereo jack cables. Their counterparts go inside the box. The white USB cable at the left goes from the Teensy to the PC.)

**Reply by 3hhh** [reply]
This is my [latest prototype](https://github.com/corrados/edrumulus/tree/main/hardware/prototype_3hhh/kicad):

![image](https://github.com/user-attachments/assets/1df1f07b-2334-46e0-bd7f-74d9047a1358)

**Comment by analogbetateam** [comment]
Thank you Volker for this wonderful project and work overall ^:}
I logged  in here just  to ask permissions if i can  upload a  Utube  video where i use your name and give credits to your work and link .
Please let me know  how/where i can share my  private link so you decide yourself if i can upload it public  or otherwise 
.My description would be something like :

Edrumulus_ Proof of Concept  .
First test on Wemos Lolin esp32 with 12 GPIO  in use, with 6 of them as double zone pads .
Old  MacBookPro with Logic X Drum Designer .
Sorry for the bad mobile 's  recording sound quality .
Thanks to Volker Fischer's  wonderful project !!  :^}    https://github.com/corrados/edrumulus

**Reply by analogbetateam** [reply]
![0-02-05-5502ce359627dfca7621ed67d789226b738150a7ca8071fd6416841174df881f_1c6dc560736fb5](https://github.com/user-attachments/assets/cea4d70b-ab27-4c38-99b8-64b1127cab81)

**Reply by corrados** [reply]
> I logged in here just to ask permissions if i can upload a Utube video

Yes, sure you can :-). No need to show me the video in advance.

**Reply by analogbetateam** [reply]
Thanks .I'm 'clueless' of code and/or how github works so i had to 'read' EVERYthing included here to come up with a working solution but i'm quite experienced in reading /understanding schematics measuring and soldering ^:} I happen to understand quite well how  signals detection  seems to work as i was deep into  analog guitar synthesizer  'waters' .
I  probably have myriad 'questions ' to ask { mostly on GUI parameters settings/use of} and must understand where/how  to do so , but for starters this  seems to work like a charm .KUDOS :
https://www.youtube.com/watch?v=lJCxvKZ8xkM

**Reply by corrados** [reply]
Great to see that you are a jazz-style drummer. I assume you play much softer than I do as a rock drummer. How do you like the sensitivity of Edrumulus?

**Reply by analogbetateam** [reply]
Well ,thanks for the compliment but certainly I m not a trained drummer in any means ^:)  
I ‘ve been pro working guitarist for almost 28 years (retired now)  and also play other instruments (oud,clarinet,sax ,bass etc) as I can enjoy, for fun  . As for the jazz style well ,I simply like the left hand traditional grip ^:) .  Your playing over the groovy backing track (prototype 1) was beyond  what I can ever do and your dexterity on the drumgizmo MMX paradidles  was impressive and spot on !! KUDOS.
   Your project is very well thought but being  ‘heavy’ multi-file code is complex and I STILL  need to understand how some things work (linked) together to get the best out of it .
   
     For the history  now:  
     I started edrumulus  after I found a full set of hard plastic  set (without brain  but with its stand plus the kick pedal ) in (believe it or not ) the recycle garbage bin !! The exact set is this : Activision wireless Drum kit controller XbOx 360. https://u-mercari-images.mercdn.net/photos/m42045908150_2.jpg .
       I tested  the pads with an oscilloscope  and all had piezo ,but  were dead mute, so needed re soldering to work .That was ‘easy’ .
       For the record (because I also use LtSpice a lot )your 100k input resistor before each GPIO is just ‘fine and safe’ IMHO ,cause these piezo I tried could easily ’spit’ a healthy … yes 22Volts AC out, when hit hard !Also polarity seems important but I didn’t bother much for it now . 
       I prototyped one channel with your values for biasing at 1.6Volt and Shotckie diodes for protection and was happy  with first results . Then I decided that I didn’t like the pad size and hard rubber  feel and got myself three nice quality DIY mesh 2ply Remo pad for snare tom and floor tom as seen here:   
       
![20251105_075918](https://github.com/user-attachments/assets/84785f1c-3860-4d84-819e-76353e523c27)

       Then I start reading all your discussions ,issues etc here and tried to understand your code’s comments by having ALL code and header/library  files in one SubEthaEdit file before I uploaded into a Wemos Lolin 32 to try. 
       For the small board prototype i use for now , I had all parts soldered  as near as possible to the GPIOs (common sense) so not visible as some are under the Lolin itself while other are  under the bottom of  board .  
       I still haven’t used capacitors at the ground side of bias pair as filtering (as discussed in your latest hardware prototypes) but for starters I can have most thresholds around 6 or 7 without false  noise triggered events and  the sensitivities as low as I like (6-9) seems to give an excellent range of play feel  overall.  
       Ofcourse I had to ‘learn’ the hard way cause once I activated more than three channels at once with floating GPIOs I had a crazy drummer play non stop solo along due to noise floors^:) 
       I struggled to make sense on how the python GUI works on my old MacBook but at the end I figured out how to ’navigate’ inside it and changes parameters .
 My DIY mesh pads (look of great quality) are NOT with a piezo at the center BUT have one piezo element under a Roland -style pyramid foam near the upper edge of the whole pad and one more near ,glued for the rimshot .
 So positional sensing is not a case here ( i believe)  although I see the M-E moving nicely on the python GUI .
  Important ‘eureka’ moment was for me that I should  start/pick a pad type as close to mine and understand the parameters and how they react . For now I mostly use    PD120 or PDX100 and for ribbon switch  Lemon cymbal
CY6 or CY5 .For Hihat pedal CC I got a 20K potentiometer from aDrum-Tec universal HiHat controller that found cheap non working  and fixed  it myself. 

 I understand that the proper way to do this is to use Octave and do the procedure (record hits and implement into code ) you explained BUT I had difficulties to understand how Octave works , once I  installed it , so for now I ‘play’ with adjusting /trying  values . 
 Besides some crosstalk issues I get ,overall edrumulus seems VERY NICELY responsive !!!
  By crosstalk I mean some heavy Lemon Crash rim switch hits activate the Snare rim midi note BUT THAT MIGHT HAS TO DO with the way I setup the hardware sitting on one big flat wooden ‘floor’ and ’tied’ together as seen here: 
  
![20251105_080007](https://github.com/user-attachments/assets/a5365594-11ee-4917-94fd-968bb0d89b73)
![20251105_080124](https://github.com/user-attachments/assets/7ad9565b-e89d-4129-9c9d-cbc5354b0b39)

  Some sort of  foam between/where pads are tighten into metalwork might reveal better isolation IMHO  and need to be tried ASAP

Overall I’m trying to built it from what spare parts  I have or can find . 
Sorry for this long answer I hope I don’t “violate’ any rules and here’s some questions if you kindly find some time:

  Although I ‘understood’ (I hope I did )  from your  GUI the cmd_names interaction of    "type (most important initial  parameter )“,  "thresh", "sens", "pos thres", "pos sens", "rim thres”(adjust for isolation ie distinguishing between a head hit and a rim shot), "mask (how long to look for peaks)”,curve","note" ”rim boost"and "note rim", 
my ‘puzzle’ is A: what ”rim/pos", 0-3 is doing  
 B: If ”cross",  means crosstalk between same pad ’s mesh VS its rim or crosstalk between  overall kit’s pads in general .  C:”note2" /note2 rim", "rim pos thres", "rim pos sens",   Do I need these parameters ?     
  D "coupling", I suspect has to do with special double piezo plus switch cymbals ? Not my case  and 
  E: "spike (GLOBAL) is an esp32 workaround due to imperfect ADC ?  
  Since I’m not using center piezo I ‘m more interested in dual zone sounds  separation as possible so I think  on how to adjust Parameters like this please correct me if i'm lost  :  
  Rim shot detection (using rim_shot_threshold) is about distinguishing between a head hit and a rim shot (or rim-only hit). Is this  what I need  to adjust for isolation?

 Rim shot positional sensing is about determining where on the rim you hit (like on a ride cymbal for bell/edge). I think I  don't need this ? and 
can ignore the rim_pos_threshold and rim_pos_sensitivity ?
 Instead I should focus on:
rim_shot_threshold: to set the level at which a rim shot is detected (relative to the head hit) and
 rim_shot_boost: if i want to boost the rim shot velocity  ?
 My test are done on Logic X Drum Designer where i had to activate the 'hidden' V-Drum input mapping tick box.
 Still some midi notes need more recearch to "obey" to what i adjust on python GUI compared to Logic"s midi notes .
 
On cymbals,  the switch is a digital on/off. Right?  
So for my DIY mesh pads case  a preset like PD120 (which has pos_sense_is_used true and rim_shot_is_used true) is ‘ok’ to  adjust parameters

Can I use threshold of 31 and sensitivity of 1  since/if  I  don't want head positional sensing?
I need to adjust rim_shot_threshold until I get good separation between head and rim right ?.

For the ribbon switch cymbal a preset like CY6  has is_rim_switch true and rim_shot_is_used true ,  seems ok ,
but rim_shot_threshold doesn't apply here because it's a switch?  I'm lost here as well as why the HihaT controller has all these parameters as well .Fortunately my HihaT pedal works ok although i can't make the volume of press foot only sound equal to other Hihat open/close levels  .  Am I close or completely ’stupid’ ? ^:)
 Last one :
Would a parallel setup of ‘for example  snare  make use of DIFFERENT PAD (part) ?
 I mean could a single snare lets say  // analog pins setup:               snare 
static int analog_pins4[]         = { 36  };
static int analog_pins_rimshot4[] = { 35  }; //  and obviously 
const int number_pads4 =1 // can have  GPIO 35 on a different pad than GPIO 36 ?

Although i'm an analog circuit DIYer i 've  successfully built an awfully complex  EWI Clarinet project with an ESP32 these last two years ( a German friend helped  me on code plus many AI chats and research  ) but your project ,,, man  is a too big 'meal to digest " ^:) 

Idea: IMHO the use of a battery (instead USB) for powering up the system theoretically speaking would make the bias reference much quieter  so less ADC spikes and cleaner supply overall .I believe a modified USB cable to interface with laptop but NOT powering up could be an option . I will try it it when free time 

   Sorry for this lengthy  response ,be always well  and kudos again .Thanks

**Reply by 3hhh** [reply]
Wow, looks like quite a journey you did there! Congrats to succeeding!

>  For the record (because I also use LtSpice a lot )your 100k input resistor before each GPIO is just ‘fine and safe’ IMHO ,cause these piezo I tried could easily ’spit’ a healthy … yes 22Volts AC out, when hit hard !

Yes, imho the correct value depends on the piezo / underlying hardware and sensitivity one wants to achieve (and in what range one wants to achieve the sensitivity, i.e. hard hitter vs soft hitter).

> A: what ”rim/pos", 0-3 is doing  

Enables/disables rim shots in combination with pos sensing IIRC.

> B: If ”cross", means crosstalk between same pad ’s mesh VS its rim or crosstalk between overall kit’s pads in general .

Don't remember. Use [my tool](https://github.com/3hhh/xtalk) for crosstalk cancellation.

> C:”note2" /note2 rim", "rim pos thres", "rim pos sens", Do I need these parameters ?  

note2 is some MIDI note that some parts may spit out (don't remember which one exactly, probably hihat open vs half open vs closed); the others are about experimental rim shot positional sensing (see the edrumulus changelog for details).

> D "coupling", I suspect has to do with special double piezo plus switch cymbals ? Not my case and

Yes.

> E: "spike (GLOBAL) is an esp32 workaround due to imperfect ADC ?  

I don't remember that one, but I don't use esp32.

> Since I’m not using center piezo I ‘m more interested in dual zone sounds separation as possible so I think on how to adjust
> Parameters like this please correct me if i'm lost : Rim shot detection (using rim_shot_threshold) is about distinguishing between a head hit and a rim shot (or rim-only hit). Is this what I need to adjust for isolation?

Yes. It's unrelated to center vs side piezo though. You can still put your piezo in the center to ideally also achieve pos sensing. Most people only need it on the snare though.
Pos sensing only uses the two parameters `pos sensing threshold` & `pos sensing sensitivity` or so.

> On cymbals, the switch is a digital on/off. Right? 

Most commercially available cymbals have a single piezo for hit velocity and a switch for rim shot detection.
The latter is mostly an on/off, although it's actual analogue characteristics are surprisingly less so, which is why you have that threshold parameter to decide when it should be considered an on vs an off.

> I'm lost here as well as why the HihaT controller has all these parameters as well .

IIRC the hihat controller is only affected by a small subset of parameters (probably threshold & sensitivity), but the UI shows the same parameter set as for all other pads.

**Reply by analogbetateam** [reply]
Thank you '3hhh' for these clarifications ^:)
I realized that edrumulus project is not 'new' but started about 5 years ago,  and that ,besides 'corrados '  ,you ,  along with 'jstma' were  one of these people who 'collaborated 'and came up with this wonderful (yet quite complex ) project .
I installed a linux distro on an old laptop just to be able to 'see' all of hardware versions on Kicad ( i mainly use Eagle) and  yours version is the Teency one .Nice clean PCB work  ^:)
I see that edrumulus is a 'mature' project while i still scratched the surface with my simplistic prototype board .
I will 'dig' deeper into studying the 'discussions' and 'issues' and change log files and i'm pretty confident that i'll finally come up with a stable setup .
Your MIDI cross-talk cancellation filter 'tool' seems great idea but maybe not an option for me ,since i'm neither on linux platform nor experienced with code/python stuff .

OTOH for an over 60 plus analog DIYer my efforts so far seem that i can still learn 'new' things ^:) 
Yesterday i moved my Lemon crash pad from its stand and realized that my idea of ''some sort of foam between/where pads are tighten into metalwork might reveal better isolation'' is indeed VALID  and need to be seriously  tried as all ghost/false snare note events  disappeared  !

I understand that Teensy might be a better option due to Esp32 quite noisy non linear AD converters but Teensy is ridiculously expensive in here .
In fact i can only find only one in here  version DEV-16771  for 50euro !! and i only have worked and own many Esp32 boards so far .
I would like to see more 'users' who built edrumulus upload a video while performing and maybe more details of all these different 'builds' .
Overall i'm very happy so far ^:)
Thank you all 
George

**Reply by 3hhh** [reply]
> Your MIDI cross-talk cancellation filter 'tool' seems great idea but maybe not an option for me ,since i'm neither on linux platform nor experienced with code/python stuff .

It should also work on Windows & Mac; I just never tested that.
Cross-talk cancellation is also configurable via json config file (it got a bit complex admittedly) or even just command line without config for more basic stuff, i.e. you shouldn't need python skills.

Happy hacking! :)

**Reply by analogbetateam** [reply]
Happy holidays and a happy new year to all .
I got me a Teensy 4.1 and i 'd like specific info on what needs to be ''changed'' if/or  needed in the latest main code , as to try a new implementation instead my {almost perfectly working } Esp32  built .
My main 'concern ' is the changes needed to have  the Teensy working with  9   pads  like my Esp 32 does .
I see on the hardware .cpp the following settings  but these seem to be for 8 pads  as is :
#ifdef TEENSYDUINO // e,t,c code 
  static int analog_pins1[]         = { A10,   A11,   A12,        A13,       A1,     A6,    A4,    A5 };
  static int analog_pins_rimshot1[] = {  A9,    -1,    A0,         -1,       A3,     A8,    A2,    A7 };
  Do i simply have to 'add' one more Teensy pin in here arbitrarily  and  maybe un -comment  some lines on the main ino ?
  
  Also do i need to 'change' main ino's 
  static int analog_pins4[]   and 
static int analog_pins_rimshot4[]  
to the above Teensy static int analog_pins1[} ?
  Thanks in advance for any help ^:}

**Comment by 3hhh** [comment]
On 12/31/25 12:39 PM, GeorgeF wrote:
> Happy holidays and a happy new year to all .

Same to you!

> I got me a Teensy 4.1 and i 'd like specific info on what needs to be ''changed'' if/or  needed in the latest main code , as to try a new implementation instead my {almost perfectly working } Esp32  built .
> My main 'concern ' is the changes needed to have  the Teensy working with  9   pads  like my Esp 32 does .
> I see on the hardware .cpp the following settings  but these seem to be for 8 pads  as is :
> #ifdef TEENSYDUINO // e,t,c code
>    static int analog_pins1[]         = { A10,   A11,   A12,        A13,       A1,     A6,    A4,    A5 };
>    static int analog_pins_rimshot1[] = {  A9,    -1,    A0,         -1,       A3,     A8,    A2,    A7 };
>    Do i simply have to 'add' one more Teensy pin in here arbitrarily  and  maybe un -comment  some lines on the main ino ?

Yes, exactly.
You also need to modify the edrumulus_gui.py to be able to modify the settings (`pad_names` array IIRC).
In my tests the Teensy 4.1 CPU maxed out at 14 used analog outputs.

**Reply by analogbetateam** [reply]
Thanks , although i could 'use' some more clarifications .
You mean that your Teensy built maxed out at 14 used analog outputs like in the initial example {8pads with 6 rimshts ie 8+6=14 } BUT if i opt for 9 pads total then i could setup 9 as long as i don't define/use   dual pads on the toms for example ?
Also the modification of edrumulus_gui.py settings (`pad_names` array ) as you state ie:
pad_names    = ["snare", "kick", "hi-hat", "ctrl", "crash", "tom1", "ride", "tom2", "tom3"]
i really 'don't get it' to need Teensy specific changes .!
Anyway thanks and Happy New Year

PS : Seems like main ino"s setup needs something like:
void setup(){
  // get the pin-to-pad assignments
  int* analog_pins = analog_pins4; // ichange to analog_pins1 []
  int* analog_pins_rimshot = analog_pins_rimshot4; // pins1 [] as above for Teensy ?

**Reply by 3hhh** [reply]
> You mean that your Teensy built maxed out at 14 used analog outputs like in the initial example {8pads with 6 rimshts ie 8+6=14 } BUT if i opt for 9 pads total then i could setup 9 as long as i don't define/use dual pads on the toms for example ?

Yes.

For example I currently use
```
static int analog_pins1[]         = { A7,  A1, A6, A0,  A2, A10,  A9, A11, A12, A13 };
static int analog_pins_rimshot1[] = { A14, -1, A4, -1,  -1,  -1,  A3,  -1, -1,  A5  };
```
and
```
pad_names    = ["snare", "kick", "hi-hat", "ctrl", "crash", "tom1", "ride", "tom2", "tom3", "china", "add2", "add3"]
```
The latter is not specific to the Teensy, you also need it for other platforms.

**Reply by 3hhh** [reply]
@corrados now also added [official support for another pad](https://github.com/corrados/edrumulus/pull/176).
